File: 卡码/代码随想录/JS/栈与队列/04-滑动窗口/LeetCode_0239.js
//239. 滑动窗口最大值
class List{
  queue;
  constructor(){
    this.queue=[];
  }
  empty(){
    return this.queue.length===0;
  }
  front(){
    return this.queue[0];
  }
  back(){
    return this.queue[this.queue.length-1];
  }
  push(value){
    while(!this.empty()&&value>this.back()) this.queue.pop();
    this.queue.push(value);
  }
  pop(value){
    if(!this.empty()&&value==this.front()) this.queue.shift();
  }
}
const maxSlidingWindow = function(nums, k) {
  const queue=new List();
  const res=new Array();
  for(let i=0;i<k;i++) queue.push(nums[i]);
  res.push(queue.front());
  for(let i=k;i<nums.length;i++){
    queue.pop(nums[i-k]);
    queue.push(nums[i]);
    res.push(queue.front());
  }
  return res;
};

File: 卡码/代码随想录/C++/栈与队列/04-滑动窗口/LeetCode_0239.cpp
//239. 滑动窗口最大值
#include <vector>
#include <deque>
using namespace std;

class List{
	public:
		int front(){
			return q.front();
		}
		void push(int val){
			while(!q.empty()&&val>q.back()) q.pop_back();
			q.push_back(val);
		}
		void pop(int val){
			if(!q.empty()&&q.front()==val) q.pop_front();
		}
	private:
		deque<int> q;
};

class Solution {
public:
  vector<int> maxSlidingWindow(vector<int>& nums, int k) {
		vector<int> res;
		List list;
		for(int i=0;i<k;i++) list.push(nums[i]);
		res.push_back(list.front());
		for(int i=k;i<nums.size();i++){
			list.pop(nums[i-k]);
			list.push(nums[i]);
			res.push_back(list.front());
		}
		return res;
  }
};
